File: src/main/java/com/github/sorusclient/client/setting/SettingManager.java
package com.github.sorusclient.client.setting;

import org.json.JSONObject;

import java.io.File;
import java.util.*;

public class SettingManager {

    private final File PROFILE_FILE = new File("sorus/profile");
    private final Map<String, SettingContainer> settingContainers = new HashMap<>();

    private Profile mainProfile;
    private Profile currentProfile;

    private final List<Profile> allProfiles = new ArrayList<>();

    public void register(SettingContainer settingContainer) {
        this.settingContainers.put(settingContainer.getId(), settingContainer);
    }

    public void loadProfiles() {
        this.mainProfile = Profile.read(new File(PROFILE_FILE, "main"));
        this.addAllProfiles(this.mainProfile);
    }

    private void addAllProfiles(Profile profile) {
        this.allProfiles.add(profile);
        this.allProfiles.addAll(profile.getChildren().values());
    }

    public void load(String profileId) {
        if (this.currentProfile != null) {
            this.save(this.currentProfile);
        }
        Profile profile = mainProfile.getProfile(profileId);
        load(profile);
    }

    public void load(Profile profile) {
        this.currentProfile = profile;

        this.loadInternal(profile);
    }

    public void delete(Profile profile) {
        if (profile == this.mainProfile) return;
        this.load(this.mainProfile);

        this.allProfiles.remove(profile);

        profile.delete();
    }

    @SuppressWarnings("unchecked")
    private void loadInternal(Profile profile) {
        if (profile.getParent() != null) {
            this.loadInternal(profile.getParent());
        }

        String profileSettings = profile.readSettings();
        JSONObject settingsJson = new JSONObject(profileSettings);

        for (Map.Entry<String, Object> jsonObject : settingsJson.toMap().entrySet()) {
            SettingContainer settingContainer = this.settingContainers.get(jsonObject.getKey());
            if (settingContainer != null) {
                Map<String, Object> jsonObject1 = (Map<String, Object>) jsonObject.getValue();

                settingContainer.setShared(profile == this.mainProfile);

                settingContainer.load(jsonObject1);
            }
        }
    }

    public List<Profile> getAllProfiles() {
        return this.allProfiles;
    }

    public Profile getCurrentProfile() {
        return currentProfile;
    }

    public void saveCurrent() {
        this.save(this.currentProfile);
    }

    public void save(String profileId) {
        Profile profile = mainProfile.getProfile(profileId);
        save(profile);
    }

    private void save(Profile profile) {
        boolean isMainProfile = profile.equals(this.mainProfile);

        Map<String, Object> settingsJsonParent = new JSONObject(this.mainProfile.readSettings()).toMap();
        Map<String, Object> settingsJson = new HashMap<>();

        for (Map.Entry<String, SettingContainer> settingContainer : this.settingContainers.entrySet()) {
            if (settingContainer.getValue().isShared() && !isMainProfile) {
                settingsJsonParent.put(settingContainer.getKey(), settingContainer.getValue().save());
            } else {
                settingsJson.put(settingContainer.getKey(), settingContainer.getValue().save());
            }
        }

        JSONObject jsonObject = new JSONObject(settingsJson);
        profile.writeSettings(jsonObject.toString(2));

        if (!isMainProfile) {
            JSONObject jsonObjectParent = new JSONObject(settingsJsonParent);
            profile.getParent().writeSettings(jsonObjectParent.toString(2));

            this.updateProfile(this.mainProfile);
        }
    }

    private void updateProfile(Profile profile) {
        JSONObject settings = new JSONObject(profile.readSettings());
        Map<String, Object> newSettings = new HashMap<>();

        for (Map.Entry<String, Object> jsonObject : settings.toMap().entrySet()) {
            SettingContainer settingContainer = this.settingContainers.get(jsonObject.getKey());
            if (profile == mainProfile || settingContainer == null || !settingContainer.isShared()) {
                newSettings.put(jsonObject.getKey(), jsonObject.getValue());
            }
        }

        profile.writeSettings(new JSONObject(newSettings).toString(2));

        for (Profile profile1 : profile.getChildren().values()) {
            this.updateProfile(profile1);
        }
    }

    public void createNewProfile() {
        List<String> currentNames = new ArrayList<>();
        for (Profile profile : this.allProfiles) {
            if (profile.getId().length() > 1) {
                currentNames.add(profile.getId().substring(1, profile.getId().length() - 1));
            }
        }

        Profile newProfile = null;
        int i = 0;
        while (newProfile == null) {
            String id = "profile" + i;
            if (!currentNames.contains(id)) {
                newProfile = this.mainProfile.createProfile("/" + id + "/");
            }
            i++;
        }

        this.allProfiles.add(newProfile);

        this.load(newProfile);
    }

}

File: src/main/kotlin/com/github/sorusclient/client/setting/SettingManager.kt
/*
 * Copyright (c) 2022. Danterus
 * Copyright (c) 2022. Sorus Contributors
 *
 * SPDX-License-Identifier: MPL-2.0
 */

package com.github.sorusclient.client.setting

import com.github.sorusclient.client.bootstrap.server.ServerIntegrationManager
import com.github.sorusclient.client.setting.data.CategoryData
import com.github.sorusclient.client.setting.display.DisplayedCategory
import org.apache.commons.io.FileUtils
import org.json.JSONObject
import java.io.File
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.HashMap

object SettingManager {

    private val profileFile = File("sorus/profile")
    lateinit var mainProfile: Profile
    var currentProfile: Profile? = null
        private set
    private val allProfiles: MutableList<Profile> = ArrayList()

    val settingsCategory: CategoryData = CategoryData()
    val mainUICategory: DisplayedCategory = DisplayedCategory("Main")

    var cachedSettings: Map<String, Any>? = null

    init {
        profileFile.mkdirs()
    }

    fun initialize() {
        loadProfiles()
        load("/")
        ServerIntegrationManager.registerJoinListener("settings", this::handleJoinForced)
        ServerIntegrationManager.registerLeaveListener(this::handleLeaveForced)
    }

    private fun handleJoinForced(json: Any) {
        settingsCategory.loadForced(json)
    }

    private fun handleLeaveForced() {
        settingsCategory.clearForced()
    }

    private fun loadProfiles() {
        val mainProfileFolder = File(profileFile, "main")
        if (!mainProfileFolder.exists()) {
            mainProfileFolder.mkdirs()
            FileUtils.write(File(mainProfileFolder, "settings.json"), "{}")
        }

        mainProfile = Profile.read(mainProfileFolder)
        addAllProfiles(mainProfile)
    }

    private fun addAllProfiles(profile: Profile) {
        allProfiles.add(profile)
        allProfiles.addAll(profile.children.values)
    }

    private fun load(profileId: String) {
        val profile = mainProfile.getProfile(profileId)!!
        load(profile)
    }

    fun load(profile: Profile) {
        currentProfile?.let { save(it) }
        currentProfile = profile
        loadInternal(profile)
    }

    fun delete(profile: Profile) {
        if (profile === mainProfile) return
        this.load(profile.parent!!)
        allProfiles.remove(profile)
        profile.delete()

        for (entry in HashMap(profile.parent!!.children)) {
            if (entry.value == profile) {
                profile.parent!!.children.remove(entry.key)
            }
        }
    }

    private fun loadInternal(profile: Profile) {
        val profiles: MutableList<Profile> = ArrayList()
        var tempProfile: Profile? = profile
        while (tempProfile != null) {
            profiles.add(0, tempProfile)
            tempProfile = tempProfile.parent
        }

        for (profileParent in profiles) {
            val map = JSONObject(profileParent.readSettings()).toMap()
            load(map, profileParent == profile)

            if (profileParent == profile) {
                cachedSettings = map
            }
        }
    }

    private fun load(json: Any, isPrimary: Boolean) {
        settingsCategory.load(json, isPrimary)
    }

    fun getAllProfiles(): List<Profile> {
        return allProfiles
    }

    fun saveCurrent() {
        save(currentProfile!!)
    }

    private fun save(profile: Profile) {
        val saved = save() as Map<*, *>

        val combined = combine(saved, cachedSettings!!)
        profile.writeSettings(JSONObject(combined as Map<*, *>).toString(2))
    }

    private fun combine(any1: Any, any2: Any): Any {
        return if (any1 is Map<*, *>) {
            val newMap = HashMap<String, Any>()

            for (thing in any2 as Map<*, *>) {
                if (any1[thing.key] != null) {
                    newMap[thing.key as String] = combine(any1[thing.key]!!, thing.value!!)
                } else {
                    newMap[thing.key as String] = thing.value!!
                }
            }

            for (thing in any1) {
                if (thing.value !is Map<*, *> || any2[thing.value!!] == null) {
                    newMap[thing.key as String] = thing.value!!
                }
            }

            newMap
        } else {
            any1
        }
    }

    private fun save(): Any {
        return settingsCategory.save()
    }

    fun createNewProfile(profile: Profile) {
        val currentNames: MutableList<String> = ArrayList()
        for (profile in profile.children.values) {
            if (profile.id.length > 1) {
                currentNames.add(profile.id.substring(1, profile.id.length - 1))
            }
        }
        var newProfile: Profile? = null
        var i = 0
        while (newProfile == null) {
            val id = "Profile$i"
            if (!currentNames.contains(id)) {
                newProfile = profile.createProfile("/$id/")
            }
            i++
        }
        allProfiles.add(newProfile)
        this.load(newProfile)
    }

}